_posts/2021-01-19-bazel-4-0.md

```diff
@@ -0,0 +1,99 @@
+* [`--incompatible_avoid_conflict_dlls=true`](https://github.com/bazelbuild/bazel/issues/11600)
+
+The following flags have either been removed or are no-ops:
+
+* [`--incompatible_load_java_rules_from_bzl`](https://github.com/bazelbuild/bazel/issues/8746) and [`--incompatible_load_proto_rules_from_bzl`](https://github.com/bazelbuild/bazel/issues/8922) are now no-ops.
+* [`--incompatible_string_replace_count`](https://github.com/bazelbuild/bazel/issues/11244) has been flipped and removed.
+* `--local_resources` and `--incompatible_remove_local_resources` [have been removed](https://github.com/bazelbuild/bazel/issues/10536). If you've been setting `--local resources` or `--incompatible_remove_local_resources=false`, migrate to using `--local_ram_resources` and `--local_cpu_resources` instead.
+* The no-op [`--deep_execroot`](https://github.com/bazelbuild/bazel/issues/12271) flag has been removed.
+
+## User Interface
+
+* Error messages emitted when an action fails have been reworked to be more informative about the failing action.
+
+## Query
+
+* Querying with `output=location` now allows the `relative_locations` flag to properly display relative locations instead of the full path. Fixes #3497.
+* The `--starlark:file` option allows the [cquery](https://docs.bazel.build/versions/master/cquery.html) feature in `--output=starlark` mode to specify the expression to format output in a file.
+
+## BEP
+
+* The BEP uses `AbortReason.OUT_OF_MEMORY` for abort events when the build tool is crashing due to OOM.
+
+## Starlark
+
+* The Starlark json module is now available. Use `json.encode(x)` to encode a Starlark value   because `JSON. struct.to_json(x)` is deprecated and will be disabled by the `--incompatible_struct_has_no_methods` flag.
+* Use `proto.encode_text(x)` to encode a Starlark value because `textproto. struct.to_proto()` is deprecated and will be disabled by the `--incompatible_struct_has_no_methods` flag. Both functions now reject list/dict fields that contain list/dict elements.
+
+## Android
+
+* Support for using AndroidX dependencies in data-binding-enabled targets
+  has been added.
+* Data-binding generation for `android_local_test` has been fixed.
+* Instrumented files for transitive sources of `android_library` and `android_binary` have been propagated.
+
+## Apple
+
+* The `relative_ast_path` feature for darwin builds to relativize swiftmodule paths for debugging has been added.
+* The `--default_ios_provisioning_profile` flag has been removed (it was a no-op).
+* `debug_prefix_map_pwd_is_dot` feature has been enabled by default on macOS. This feature passes `-fdebug-prefix-map=$PWD=.` for every compile to remove absolute paths from debug information.
+
+## Process Management / Sandboxing
```

Comment:
Do you want to add a line item for your new sandboxing page?

Reply:
Alternatively, do you want to add a section for Documentation at the bottom and include links to your versioning page, the sandboxing page, and the work on the persistent workers section? :)

Reply:
Added a Documentation section.